torch/csrc/api/src/nn/modules/activation.cpp

```diff
@@ -390,5 +391,119 @@ void ThresholdImpl::pretty_print(std::ostream& stream) const {
   stream << ")";
 }
 
+// ============================================================================
+
+MultiheadAttentionImpl::MultiheadAttentionImpl(const MultiheadAttentionOptions& options_)
+    : Module("torch::nn::MultiheadAttention"), options(options_) {
+  reset();
+}
+
+std::tuple<Tensor, Tensor> MultiheadAttentionImpl::forward(
+  const Tensor& query, const Tensor& key,
+  const Tensor& value, const Tensor& key_padding_mask,
+  bool need_weights, const Tensor& attn_mask) {
+  if (!_qkv_same_embed_dim) {
+    return F::multi_head_attention_forward(
+      query, key, value, 
+      F::MultiheadAttentionForwardFuncOptions(
+        /*embed_dim_to_check=*/options.embed_dim(),
+        /*num_heads=*/options.num_heads(),
+        /*in_proj_weight=*/in_proj_weight,
+        /*in_proj_bias=*/in_proj_bias,
+        /*bias_k=*/bias_k,
+        /*bias_v=*/bias_v,
+        /*add_zero_attn=*/options.add_zero_attn(),
+        /*dropout_p=*/options.dropout(),
+        /*out_proj_weight=*/out_proj->weight,
+        /*out_proj_bias=*/out_proj->bias
+      ).training(is_training())
+       .key_padding_mask(key_padding_mask)
+       .need_weights(need_weights)
+       .attn_mask(attn_mask)
+       .use_separate_proj_weight(true)
+       .q_proj_weight(q_proj_weight)
+       .k_proj_weight(k_proj_weight)
+       .v_proj_weight(v_proj_weight)
+    );
+  } else {
+    return F::multi_head_attention_forward(
+      query, key, value, 
+      F::MultiheadAttentionForwardFuncOptions(
+        /*embed_dim_to_check=*/options.embed_dim(),
+        /*num_heads=*/options.num_heads(),
+        /*in_proj_weight=*/in_proj_weight,
+        /*in_proj_bias=*/in_proj_bias,
+        /*bias_k=*/bias_k,
+        /*bias_v=*/bias_v,
+        /*add_zero_attn=*/options.add_zero_attn(),
+        /*dropout_p=*/options.dropout(),
+        /*out_proj_weight=*/out_proj->weight,
+        /*out_proj_bias=*/out_proj->bias
+      ).training(is_training())
+       .key_padding_mask(key_padding_mask)
+       .need_weights(need_weights)
+       .attn_mask(attn_mask)
+    );
+  }
+}
+
+void MultiheadAttentionImpl::reset() {
+  _qkv_same_embed_dim = options.kdim() == options.embed_dim() &&
```

Comment:
and after making kdim and vdim optional in options/activation.h, we can add code corresponding to these lines in activation.py
self.kdim = kdim if kdim is not None else embed_dim
self.vdim = vdim if vdim is not None else embed_dim
``if(options.kdim() != c10::nullopt) {
      options.kdim(options.embed_dim());
    }``
``if(options.vdim() != c10::nullopt) {       
options.vdim(options.embed_dim());
}``